client/human/GGHumanClientApp.cpp

```diff
@@ -1548,7 +1555,14 @@ void GGHumanClientApp::UpdateFPSLimit() {
 void GGHumanClientApp::HandleResoureDirChange() {
     if (!m_game_started) {
         DebugLogger() << "Resource directory changed.  Reparsing universe ...";
-        StartBackgroundParsing();
+        std::promise<void> barrier;
+        std::future<void> barrier_future = barrier.get_future();
+        std::thread background([this] (auto b) {
+            DebugLogger() << "Started background parser thread";
+            StartBackgroundParsing(std::move(b));
+        }, std::move(barrier));
+        background.detach();
+        barrier_future.wait();
```

Comment:
I'm confused... What's the point of creating the `promise` and `future` if you `wait()` on the future immediately after creating the thread and launching the parsing? Why not just do the parse right in this function / thread?

Reply:
`promise` and `future` got resolved right after first `"common"` parsing so it doesn't have to wait full parsing in UI thread.

Reply:
That doesn't help my confusion.

In this section of code, a thread is launched to parse, then it waits on the future, which should block this thread until the future is available.

Is this not the UI thread?

Or what other thread is doing what else in parallel during that wait?

Reply:
I guess the future is unblocked when `parse::named_value_refs` finishes, but the thread continues and launches the other asynchronous parsers after that?